src/headless/yare_impl/headless.rs

```diff
@@ -1,153 +1,622 @@
-            // and apply it after all of the commands process
             for command in player_commands.iter() {
                 match command {
                     Command::Energize { index, target } => {
                         let source_spirit = &spirits[*index];
                         let target_spirit = &spirits[*target];
-                        if !source_spirit.hp > 0 || player.index != source_spirit.player_id {
+                        if source_spirit.hp < 1
+                            || player.index != source_spirit.player_id
+                            || source_spirit.pos.dist(target_spirit.pos) > ENERGIZE_RANGE
+                        {
+                            // dbg!(source_spirit);
                             continue;
                         }
                         // self energize
                         if index == target {
-                            for star in &self.stars {
+                            for (i, star) in self.stars.iter().enumerate() {
                                 // check distance
+                                if star.pos.dist(source_spirit.pos) <= ENERGIZE_RANGE {
+                                    // dbg!(source_spirit);
+                                    self.charging_spirits[i].push(*index);
+                                }
                             }
                         } else {
                             // charge friendly
+                            if self.replay_path.is_some() {
+                                energizes.push(ReplayEnergize::spirit_to_spirit(
+                                    source_spirit,
+                                    target_spirit,
+                                ));
+                            }
+                            self.spirit_energy_changes[*index] -= source_spirit.energize_amount();
                             if source_spirit.player_id == target_spirit.player_id {
-
-                                // attack
+                                self.spirit_energy_changes[*target] +=
+                                    source_spirit.energize_amount();
                             } else {
+                                // attack
+                                self.spirit_energy_changes[*target] -=
+                                    2 * source_spirit.energize_amount();
+                            }
+                        }
+                    }
+                    Command::EnergizeBase { index, target } => {
+                        let source_spirit = &spirits[*index];
+                        let target_base = &bases[*target];
+                        if source_spirit.hp < 1
+                            || player.index != source_spirit.player_id
+                            || source_spirit.pos.dist(target_base.pos) > ENERGIZE_RANGE
+                        {
+                            // dbg!(source_spirit);
+                            continue;
+                        }
```

Comment:
You repeat this section of code once for every type of energize command. Maybe put it in a function?